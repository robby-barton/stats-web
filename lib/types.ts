/* istanbul ignore file */

export type YearRanks = {
	weeks: number;
	postseason: boolean;
};

export type AvailRanks = {
	[index: string]: YearRanks;
};

export type Team = {
	team_id: number;
	name: string;
	logo: string;
	logo_dark: string;
};

export type AvailTeams = {
	[index: string]: Team;
};

export type Rank = {
	team: Team;
	conf: string;
	record: string;
	srs_rank: number;
	sos_rank: number;
	final_raw: number;
	final_rank: number;
};

export type TeamRank = {
	team: Team;
	final_rank: number;
	year: number;
	week: string;
	postseason: number;
};

export type TeamGames = {
	team: Team;
	sun: number;
	mon: number;
	tue: number;
	wed: number;
	thu: number;
	fri: number;
	sat: number;
	total: number;
};

export type RankingPath = {
	division: string;
	year: string;
	week: string;
};

export type TeamPath = {
	team: string;
};

export type ChartPoint = {
	week: string;
	rank: number;
	fillLevel: number;
};
