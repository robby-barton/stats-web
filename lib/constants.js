export const DIVISIONS = ['fbs', 'fcs']
