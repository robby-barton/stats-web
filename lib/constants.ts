/* istanbul ignore file */

export const SITE_TITLE = "Robby's Ranking";

export const DIVISIONS: string[] = ['fbs', 'fcs'];

export const CHART_MAX_Y = 150;

export const ERROR_IMAGES = ['/aspen.png', '/major.png', '/mona.png'];
