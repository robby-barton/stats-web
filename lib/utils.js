import sql from './db';
import { DIVISIONS } from './constants';

// need to tell it how to handle BigInts from Postgres
BigInt.prototype.toJSON = function () {
  return Number(this)
}

let rankings = null
let rankingsExpire = null
export async function availableRankings() {
  let now = Math.floor(new Date().getTime() / 1000)
  if (!rankings || rankingsExpire < now) {
    const rankingObjects = await sql`
      select 
        year, 
        max(case when postseason = 0 then week else 0 end) as week,
        max(postseason) as postseason 
      from team_week_results
      group by 
        year 
      order by 
        year desc
    `
    console.log("re-pull rankings per year")
    rankings = {}
    for (let i = 0; i < rankingObjects.length; i++) {
      const obj = rankingObjects[i]
      rankings[obj.year] = {
        weeks: obj.week,
        postseason: obj.postseason === 1 ? true : false,
      }
    }
    rankingsExpire = now + 300 // refresh every 5 minutes
  }

  return rankings
}

export async function checkRanking(division, year, week) {
  const avail = await availableRankings()
  if (DIVISIONS.includes(division.toLowerCase()) && year in avail) {
    if (week.toLowerCase() === 'final' && avail[year].postseason) {
      return true
    }
    if (week > 0 && week <= avail[year].weeks) {
      return true
    }
  }

  return false
}

export async function getRanking(fbs, year, week) {
  const opts = {
    year: Number(year),
    fbs: fbs,
  }
  if (week.toLowerCase() === "final") {
    opts.postseason = Number(1)
  } else {
    opts.postseason = Number(0)
    opts.week = Number(week)
  }

  const results = await sql`
    select 
      *
    from team_week_results
    where
      fbs = ${fbs} and
      year = ${year} and
      ${
        week.toLowerCase() === "final"
          ? sql`postseason = 1`
          : sql`week = ${week} and postseason = 0`
      }
    order by
      final_rank
  `

  return results
}

export async function getTeamRankings(team) {
  const results = await sql`
    select
      team_id,
      name,
      final_rank,
      year,
      week
    from team_week_results
    where
      team_id = ${team}
    order by
      year,
      postseason,
      week
  `
  return results
}

export async function getUniqueTeams() {
  const results = await sql`
    select 
      team_id, 
      name 
    from team_names 
    where 
      team_id in (
        select 
          distinct team_id 
        from team_week_results
      )
    order by 
      name
  `
  return results
}

export async function allGames() {
  const results = await sql`
    with gamesDOW as (
      with gamesList as (
        (
          select
            home_id as team_id,
            extract(dow from start_time) as dow,
            game_id
          from games
        ) union all (
          select
            away_id as team_id,
            extract(dow from start_time) as dow,
            game_id
          from games
        )
      )
      select
        team_id,
        sum(case when dow = 0 then 1 else 0 end) as sun,
        sum(case when dow = 1 then 1 else 0 end) as mon,
        sum(case when dow = 2 then 1 else 0 end) as tue,
        sum(case when dow = 3 then 1 else 0 end) as wed,
        sum(case when dow = 4 then 1 else 0 end) as thu,
        sum(case when dow = 5 then 1 else 0 end) as fri,
        sum(case when dow = 6 then 1 else 0 end) as sat,
        count(1) as total
      from gamesList
      group by
        team_id
    )
    select
      n.name,
      g.*
    from team_names n
      join gamesDOW g on (n.team_id = g.team_id)
    order by
      total desc
  `

  return results
}

export async function getRankingPathParams() {
  const avail = await availableRankings()
  const paths = []
  DIVISIONS.map(division => (
    Object.entries(avail).forEach(entry => {
      const [ year, value ] = entry
      const { weeks, postseason } = value
      for (let i = 1; i <= weeks; i++) {
        paths.push({params: { division: division, year: year.toString(), week: i.toString() }})
      }
      if (postseason) {
        paths.push({params: { division: division, year: year.toString(), week: 'final' }})
      }
    })
  ))

  return paths
}

export async function getTeamPathParams() {
  const results = await getUniqueTeams()
  const paths = results.map(team => ({ params: { team: team.team_id.toString() }}))

  return paths
}
