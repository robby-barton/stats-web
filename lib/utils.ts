import { DIVISIONS } from "@lib/constants";
import {
	allGamesDB,
	availableRankingsDB,
	availableTeamsDB,
	getRankedTeamsDB,
	getRankingDB,
	getTeamRankingsDB,
} from "@lib/dbFuncs";
import { AvailRanks, AvailTeams, Rank, RankingPathParams, Team, TeamGames, TeamPathParams, TeamRank } from "@lib/types";

let rankings: AvailRanks = {};
let rankingsExpire = -1;
export async function availableRankings(): Promise<AvailRanks> {
	const now: number = Math.floor(new Date().getTime() / 1000);
	if (!rankings || rankingsExpire < now) {
		const rankingObjects = await availableRankingsDB();
		if (!rankingObjects.length) {
			throw new Error("Not found");
		}

		rankings = {};
		for (let i = 0; i < rankingObjects.length; i++) {
			const obj = rankingObjects[i];
			rankings[obj.year] = {
				weeks: obj.weeks,
				postseason: obj.postseason === 1 ? true : false,
			};
		}
		rankingsExpire = now + 300; // refresh every 5 minutes
	}

	return rankings;
}

let teamInfo: AvailTeams = {};
let teamInfoExpire = -1;
export async function availableTeams(): Promise<AvailTeams> {
	const now: number = Math.floor(new Date().getTime() / 1000);
	if (!teamInfo || teamInfoExpire < now) {
		const teamInfoObjects = await availableTeamsDB();
		if (!teamInfoObjects.length) {
			throw new Error("Not found");
		}

		teamInfo = {};
		for (let i = 0; i < teamInfoObjects.length; i++) {
			const obj = teamInfoObjects[i];
			teamInfo[obj.team_id.toString()] = {
				team_id: obj.team_id,
				name: obj.name,
				logo: obj.logo || "",
				logo_dark: obj.logo_dark || "",
			};
		}
		teamInfoExpire = now + 300; // refresh every 5 minutes
	}

	return teamInfo;
}

export async function getRanking(fbs: boolean, year: number, week: string): Promise<Rank[]> {
	const results = await getRankingDB(fbs, year, week);
	const availTeams = await availableTeams();

	const data: Rank[] = [];
	for (let i = 0; i < results.length; i++) {
		data.push({
			team: availTeams[results[i].team_id.toString()],
			final_rank: results[i].final_rank,
			conf: results[i].conf,
			record:
				results[i].ties === 0
					? results[i].wins + "-" + results[i].losses
					: results[i].wins + "-" + results[i].losses + "-" + results[i].ties,
			srs_rank: results[i].srs_rank,
			sos_rank: results[i].sos_rank,
			final_raw: results[i].final_raw,
		});
	}

	return data;
}

export async function getTeamRankings(team: number): Promise<TeamRank[]> {
	const rankings = await getTeamRankingsDB(team);
	const teams = await availableTeams();

	const ranks: TeamRank[] = rankings.map((rank) => ({
		team: teams[rank.team_id.toString()],
		final_rank: rank.final_rank,
		year: rank.year,
		week: rank.week,
		postseason: rank.postseason,
	}));

	return ranks;
}

export async function getRankedTeams(): Promise<Team[]> {
	const rankedTeams = await getRankedTeamsDB();
	const allTeams = await availableTeams();

	const teams = rankedTeams.map((team) => allTeams[team.team_id.toString()]);

	return teams;
}

export async function allGames(): Promise<TeamGames[]> {
	const games = await allGamesDB();
	const teams = await availableTeams();

	const allGames: TeamGames[] = games
		.filter((gameSet) => {
			return teams[gameSet.team_id.toString()] ? true : false;
		})
		.map((gameSet) => ({
			team: teams[gameSet.team_id.toString()],
			sun: gameSet.sun,
			mon: gameSet.mon,
			tue: gameSet.tue,
			wed: gameSet.wed,
			thu: gameSet.thu,
			fri: gameSet.fri,
			sat: gameSet.sat,
			total: gameSet.total,
		}));

	return allGames;
}

export async function getRankingPathParams(): Promise<RankingPathParams[]> {
	const avail: AvailRanks = await availableRankings();
	const paths: RankingPathParams[] = [];
	DIVISIONS.map((division) =>
		Object.entries(avail).forEach((entry) => {
			const [year, value] = entry;
			const { weeks, postseason } = value;
			for (let i = 1; i <= weeks; i++) {
				paths.push({
					params: {
						division: division,
						year: year,
						week: i.toString(),
					},
				});
			}
			if (postseason) {
				paths.push({
					params: { division: division, year: year, week: "final" },
				});
			}
		})
	);

	return paths;
}

export async function getTeamPathParams(): Promise<TeamPathParams[]> {
	const rankedTeams = await getRankedTeamsDB();
	const paths = rankedTeams.map((team) => ({
		params: {
			team: team.team_id.toString(),
		},
	}));

	return paths;
}
