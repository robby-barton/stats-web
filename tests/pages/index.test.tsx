import { render } from "@testing-library/react";

import "@testing-library/jest-dom";

import { REVALIDATE } from "@lib/constants";
import { AvailRanks, Rank } from "@lib/types";
import { availableRankings, getRanking } from "@lib/utils";
import Home, { getStaticProps } from "@pages/index";

jest.mock("next/router", () => ({
	useRouter: jest.fn(),
}));

jest.mock("@lib/utils", () => ({
	__esModule: true,
	availableRankings: jest.fn(),
	getRanking: jest.fn(),
}));

afterAll(() => {
	jest.unmock("@lib/utils");
	jest.unmock("next/router");
});

afterEach(() => {
	jest.clearAllMocks();
});

beforeEach(() => {
	document.body.dataset.theme = "light";
});

describe("Home page", () => {
	it("renders correctly", async () => {
		const avail: AvailRanks = {
			"2023": {
				weeks: 16,
				postseason: true,
			},
		};
		const ranks: Rank[] = [
			{
				team_id: 2579,
				name: "South Carolina",
				conf: "SEC",
				record: "8-5",
				sos_rank: 13,
				srs_rank: 15,
				final_raw: 0.123456789,
				final_rank: 26,
			},
		];
		const { baseElement } = render(<Home availRanks={avail} fbs={ranks} week="Final" year={2023} />);

		expect(baseElement).toMatchSnapshot();
	});

	describe("getStaticProps", () => {
		it("returns correctly midyear", async () => {
			const availMock: AvailRanks = {
				"2022": {
					weeks: 10,
					postseason: false,
				},
			};
			(availableRankings as jest.Mock).mockReturnValue(availMock);

			const rankingMock: Rank[] = [
				{
					team_id: 2579,
					name: "South Carolina",
					conf: "SEC",
					final_rank: 1,
					final_raw: 0.123456789,
					record: "8-5",
					sos_rank: 2,
					srs_rank: 3,
				},
			];
			(getRanking as jest.Mock).mockReturnValue(rankingMock);

			const expected = {
				props: {
					availRanks: availMock,
					fbs: rankingMock,
					year: 2022,
					week: "10",
				},
				revalidate: REVALIDATE,
			};

			const result = await getStaticProps();

			expect(result).toEqual(expected);

			expect(availableRankings).toBeCalledTimes(1);
			expect(getRanking).toBeCalledTimes(1);
			expect(getRanking).toBeCalledWith(true, 2022, "10");
		});

		it("returns correctly final", async () => {
			const availMock: AvailRanks = {
				"2022": {
					weeks: 16,
					postseason: true,
				},
			};
			(availableRankings as jest.Mock).mockReturnValue(availMock);

			const rankingMock: Rank[] = [
				{
					team_id: 2579,
					name: "South Carolina",
					conf: "SEC",
					final_rank: 1,
					final_raw: 0.123456789,
					record: "8-5",
					sos_rank: 2,
					srs_rank: 3,
				},
			];
			(getRanking as jest.Mock).mockReturnValue(rankingMock);

			const expected = {
				props: {
					availRanks: availMock,
					fbs: rankingMock,
					year: 2022,
					week: "final",
				},
				revalidate: REVALIDATE,
			};

			const result = await getStaticProps();

			expect(result).toEqual(expected);

			expect(availableRankings).toBeCalledTimes(1);
			expect(getRanking).toBeCalledTimes(1);
			expect(getRanking).toBeCalledWith(true, 2022, "final");
		});
	});
});
