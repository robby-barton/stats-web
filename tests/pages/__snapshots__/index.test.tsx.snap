// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page renders correctly 1`] = `
<body
  data-theme="light"
>
  <div>
    <div
      class="container"
    >
      <div
        class="header"
      >
        <div
          class="headerTitle"
        >
          <a
            href="/"
          >
            Robby's Ranking
          </a>
        </div>
        <div
          class="headerLink"
        >
          <a
            href="/game-count"
          >
            <span
              class="headingMd"
            >
              Game Count
            </span>
          </a>
        </div>
        <div
          class="headerLink"
        >
          <a
            href="/teams"
          >
            <span
              class="headingMd"
            >
              Teams
            </span>
          </a>
        </div>
        <div
          class="headerLink"
        >
          <a
            href="/about"
          >
            <span
              class="headingMd"
            >
              About
            </span>
          </a>
        </div>
        <div
          class="headerToggle"
        >
          <button
            aria-label="Change to dark mode"
            class="css-1f6jnp0"
            title="Change to dark mode"
            type="button"
          >
            <span
              class="css-194expn"
            />
            <span>
              üåô
            </span>
            <span>
              ‚òÄÔ∏è
            </span>
          </button>
        </div>
      </div>
      <div
        class="contentWrap"
      >
        <div>
          <div
            class="inputArea"
          >
            <div
              class="selectorStyling"
            >
              <select
                class="divisionDropdown"
                title="division"
              >
                <option
                  value="fbs"
                >
                  FBS
                </option>
                <option
                  value="fcs"
                >
                  FCS
                </option>
              </select>
              <select
                class="yearDropdown"
                title="year"
              >
                <option
                  value="2023"
                >
                  2023
                </option>
              </select>
              <select
                class="weekDropdown"
                title="week"
              >
                <option
                  value="1"
                >
                  Week 1
                </option>
                <option
                  value="2"
                >
                  Week 2
                </option>
                <option
                  value="3"
                >
                  Week 3
                </option>
                <option
                  value="4"
                >
                  Week 4
                </option>
                <option
                  value="5"
                >
                  Week 5
                </option>
                <option
                  value="6"
                >
                  Week 6
                </option>
                <option
                  value="7"
                >
                  Week 7
                </option>
                <option
                  value="8"
                >
                  Week 8
                </option>
                <option
                  value="9"
                >
                  Week 9
                </option>
                <option
                  value="10"
                >
                  Week 10
                </option>
                <option
                  value="11"
                >
                  Week 11
                </option>
                <option
                  value="12"
                >
                  Week 12
                </option>
                <option
                  value="13"
                >
                  Week 13
                </option>
                <option
                  value="14"
                >
                  Week 14
                </option>
                <option
                  value="15"
                >
                  Week 15
                </option>
                <option
                  value="16"
                >
                  Week 16
                </option>
                <option
                  value="final"
                >
                  Final
                </option>
              </select>
            </div>
            <div
              class="searchDiv"
            >
              <input
                class="searchInput"
                placeholder="Search Teams"
                type="search"
              />
            </div>
          </div>
          <table>
            <thead>
              <tr>
                <th
                  class=""
                  colspan="1"
                >
                  <div
                    class="cursor-pointer select-none"
                  >
                    Rank
                  </div>
                </th>
                <th
                  class=""
                  colspan="1"
                >
                  <div
                    class="cursor-pointer select-none"
                  >
                    Team
                  </div>
                </th>
                <th
                  class=""
                  colspan="1"
                >
                  <div
                    class="cursor-pointer select-none"
                  >
                    Conf
                  </div>
                </th>
                <th
                  class=""
                  colspan="1"
                >
                  <div
                    class=""
                  >
                    Record
                  </div>
                </th>
                <th
                  class=""
                  colspan="1"
                >
                  <div
                    class="cursor-pointer select-none"
                  >
                    SRS
                  </div>
                </th>
                <th
                  class=""
                  colspan="1"
                >
                  <div
                    class="cursor-pointer select-none"
                  >
                    SOS
                  </div>
                </th>
                <th
                  class="lastColumn"
                  colspan="1"
                >
                  <div
                    class="cursor-pointer select-none"
                  >
                    Final
                  </div>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td
                  class=""
                >
                  26
                </td>
                <td
                  class=""
                >
                  <div
                    class="container"
                  >
                    <div
                      class="logo"
                    >
                      <img
                        alt="South Carolina"
                        data-nimg="fill"
                        decoding="async"
                        loading="lazy"
                        sizes="100vw"
                        src="http://localhost/_next/image?url=%2Flogos%2Flight%2F2579.png&w=3840&q=75"
                        srcset="/_next/image?url=%2Flogos%2Flight%2F2579.png&w=640&q=75 640w, /_next/image?url=%2Flogos%2Flight%2F2579.png&w=750&q=75 750w, /_next/image?url=%2Flogos%2Flight%2F2579.png&w=828&q=75 828w, /_next/image?url=%2Flogos%2Flight%2F2579.png&w=1080&q=75 1080w, /_next/image?url=%2Flogos%2Flight%2F2579.png&w=1200&q=75 1200w, /_next/image?url=%2Flogos%2Flight%2F2579.png&w=1920&q=75 1920w, /_next/image?url=%2Flogos%2Flight%2F2579.png&w=2048&q=75 2048w, /_next/image?url=%2Flogos%2Flight%2F2579.png&w=3840&q=75 3840w"
                        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
                      />
                    </div>
                    <div
                      class="name"
                    >
                      <span>
                        South Carolina
                      </span>
                    </div>
                  </div>
                </td>
                <td
                  class=""
                >
                  SEC
                </td>
                <td
                  class=""
                >
                  8-5
                </td>
                <td
                  class=""
                >
                  15
                </td>
                <td
                  class=""
                >
                  13
                </td>
                <td
                  class="lastColumn"
                >
                  0.12346
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <div
        class="footer"
      >
        <a
          href="https://github.com/robby-barton/stats-web"
        >
          &lt;/&gt; - Site source
        </a>
        <br />
        <a
          href="https://github.com/robby-barton/stats-go"
        >
          &lt;/&gt; - Ranking source
        </a>
      </div>
    </div>
  </div>
</body>
`;
